File: sdks/python/hatchet_sdk/worker/durable_eviction/cache.py
from __future__ import annotations

from datetime import datetime, timedelta
from enum import Enum

try:
    from typing import assert_never
except ImportError:
    from typing_extensions import assert_never

from pydantic import BaseModel

from hatchet_sdk.logger import logger
from hatchet_sdk.runnables.action import ActionKey
from hatchet_sdk.runnables.eviction import EvictionPolicy


class EvictionCause(str, Enum):
    TTL_EXCEEDED = "ttl_exceeded"
    CAPACITY_PRESSURE = "capacity_pressure"
    WORKER_SHUTDOWN = "worker_shutdown"


class DurableRunRecord(BaseModel):
    key: ActionKey
    step_run_id: str
    eviction_policy: EvictionPolicy | None
    registered_at: datetime

    # Waiting state
    waiting_since: datetime | None = None
    wait_kind: str | None = None
    wait_resource_id: str | None = None

    # Set by the eviction manager before requesting eviction
    eviction_reason: str | None = None

    @property
    def is_waiting(self) -> bool:
        return self.waiting_since is not None


class DurableEvictionCache:
    def __init__(self) -> None:
        self._runs: dict[ActionKey, DurableRunRecord] = {}

    def register_run(
        self,
        key: ActionKey,
        step_run_id: str,
        now: datetime,
        eviction_policy: EvictionPolicy | None,
    ) -> None:
        self._runs[key] = DurableRunRecord(
            key=key,
            step_run_id=step_run_id,
            eviction_policy=eviction_policy,
            registered_at=now,
        )

    def unregister_run(self, key: ActionKey) -> None:
        self._runs.pop(key, None)

    def get(self, key: ActionKey) -> DurableRunRecord | None:
        return self._runs.get(key)

    def get_all_waiting(self) -> list[DurableRunRecord]:
        return [r for r in self._runs.values() if r.is_waiting]

    def mark_waiting(
        self,
        key: ActionKey,
        now: datetime,
        wait_kind: str,
        resource_id: str,
    ) -> None:
        rec = self._runs.get(key)
        if not rec:
            return

        rec.waiting_since = now
        rec.wait_kind = wait_kind
        rec.wait_resource_id = resource_id

    def mark_active(self, key: ActionKey, now: datetime) -> None:
        rec = self._runs.get(key)
        if not rec:
            return

        # Clear waiting state
        rec.waiting_since = None
        rec.wait_kind = None
        rec.wait_resource_id = None

    def _capacity_pressure(
        self, durable_slots: int, reserve_slots: int, waiting_count: int
    ) -> bool:
        if durable_slots <= 0:
            return False

        max_waiting = durable_slots - reserve_slots
        if max_waiting <= 0:
            return False

        return waiting_count >= max_waiting

    def select_eviction_candidate(
        self,
        now: datetime,
        durable_slots: int,
        reserve_slots: int,
        min_wait_for_capacity_eviction: timedelta,
    ) -> ActionKey | None:
        waiting: list[DurableRunRecord] = [
            r
            for r in self._runs.values()
            if r.is_waiting and r.eviction_policy is not None
        ]

        if not waiting:
            return None

        # Prefer TTL-eligible candidates first.
        ttl_eligible: list[DurableRunRecord] = [
            r
            for r in waiting
            if r.eviction_policy is not None
            and r.eviction_policy.ttl is not None
            and r.waiting_since is not None
            and (now - r.waiting_since) >= r.eviction_policy.ttl
        ]

        if ttl_eligible:
            ttl_eligible.sort(
                key=lambda r: (
                    r.eviction_policy.priority if r.eviction_policy else 0,
                    r.waiting_since or now,
                )
            )
            chosen = ttl_eligible[0]
            ttl = chosen.eviction_policy.ttl if chosen.eviction_policy else None
            chosen.eviction_reason = _build_eviction_reason(
                EvictionCause.TTL_EXCEEDED, chosen, ttl=ttl
            )
            logger.debug(
                "DurableEvictionCache: TTL eviction candidate selected "
                f"step_run_id={chosen.step_run_id} kind={chosen.wait_kind}"
            )
            return chosen.key

        # Capacity eviction: only if we're above waiting capacity and run allows it.
        capacity_pressure = self._capacity_pressure(
            durable_slots=durable_slots,
            reserve_slots=reserve_slots,
            waiting_count=len(waiting),
        )
        if not capacity_pressure:
            return None

        capacity_candidates: list[DurableRunRecord] = [
            r
            for r in waiting
            if r.eviction_policy
            and r.eviction_policy.allow_capacity_eviction
            and r.waiting_since is not None
            and (now - r.waiting_since) >= min_wait_for_capacity_eviction
        ]

        if not capacity_candidates:
            return None

        capacity_candidates.sort(
            key=lambda r: (
                r.eviction_policy.priority if r.eviction_policy else 0,
                r.waiting_since or now,
            )
        )
        chosen = capacity_candidates[0]
        chosen.eviction_reason = _build_eviction_reason(
            EvictionCause.CAPACITY_PRESSURE, chosen
        )
        logger.debug(
            "DurableEvictionCache: capacity eviction candidate selected "
            f"step_run_id={chosen.step_run_id} kind={chosen.wait_kind}"
        )
        return chosen.key


def _build_eviction_reason(
    cause: EvictionCause,
    rec: DurableRunRecord,
    ttl: timedelta | None = None,
) -> str:
    wait_desc = rec.wait_kind or "unknown"
    if rec.wait_resource_id:
        wait_desc = f"{wait_desc}({rec.wait_resource_id})"

    match cause:
        case EvictionCause.TTL_EXCEEDED:
            ttl_str = f" ({ttl})" if ttl else ""
            return f"Wait TTL{ttl_str} exceeded while waiting on {wait_desc}"
        case EvictionCause.CAPACITY_PRESSURE:
            return f"Worker at capacity while waiting on {wait_desc}"
        case EvictionCause.WORKER_SHUTDOWN:
            return f"Worker shutdown while waiting on {wait_desc}"
        case _ as unreachable:
            assert_never(unreachable)

File: sdks/ruby/src/lib/hatchet/worker/durable_eviction/cache.rb
# frozen_string_literal: true

require "monitor"

module Hatchet
  module WorkerRuntime
    module DurableEviction
      # Eviction causes produced by :class:`DurableEvictionCache`.
      module EvictionCause
        TTL_EXCEEDED = :ttl_exceeded
        CAPACITY_PRESSURE = :capacity_pressure
        WORKER_SHUTDOWN = :worker_shutdown
      end

      # Per-run state tracked by the cache.
      #
      # ``wait_count`` is ref-counted so concurrent waits over the same durable
      # run don't prematurely clear the waiting flag when one child completes
      # before the others.
      class DurableRunRecord
        attr_reader :key, :step_run_id, :invocation_count, :eviction_policy, :registered_at
        attr_accessor :waiting_since, :wait_kind, :wait_resource_id, :eviction_reason, :wait_count

        # @param key [String] The action key uniquely identifying this step run invocation.
        # @param step_run_id [String]
        # @param invocation_count [Integer]
        # @param eviction_policy [Hatchet::EvictionPolicy, nil]
        # @param registered_at [Time]
        def initialize(key:, step_run_id:, invocation_count:, eviction_policy:, registered_at:)
          @key = key
          @step_run_id = step_run_id
          @invocation_count = invocation_count
          @eviction_policy = eviction_policy
          @registered_at = registered_at
          @waiting_since = nil
          @wait_kind = nil
          @wait_resource_id = nil
          @eviction_reason = nil
          @wait_count = 0
        end

        def waiting?
          @wait_count.positive?
        end
      end

      # Thread-safe in-memory cache of waiting durable task invocations.
      #
      # Mirrors :class:`hatchet_sdk.worker.durable_eviction.cache.DurableEvictionCache`
      # from the Python SDK. All public methods lock an internal monitor.
      class DurableEvictionCache
        def initialize
          @runs = {}
          @monitor = Monitor.new
        end

        # Register a new durable run invocation.
        def register_run(key, step_run_id:, invocation_count:, now:, eviction_policy:)
          @monitor.synchronize do
            @runs[key] = DurableRunRecord.new(
              key: key,
              step_run_id: step_run_id,
              invocation_count: invocation_count,
              eviction_policy: eviction_policy,
              registered_at: now,
            )
          end
        end

        # Unregister a durable run invocation.
        def unregister_run(key)
          @monitor.synchronize { @runs.delete(key) }
        end

        # Fetch the record for a given key.
        # @return [DurableRunRecord, nil]
        def get(key)
          @monitor.synchronize { @runs[key] }
        end

        # @return [Array<DurableRunRecord>]
        def all_waiting
          @monitor.synchronize { @runs.values.select(&:waiting?) }
        end

        # @param step_run_id [String]
        # @return [String, nil] the action key for the matching record
        def find_key_by_step_run_id(step_run_id)
          @monitor.synchronize do
            @runs.each do |k, rec|
              return k if rec.step_run_id == step_run_id
            end
            nil
          end
        end

        # Mark the run as waiting (ref-counted). Increments the wait counter and
        # stores the wait metadata on the record.
        def mark_waiting(key, now:, wait_kind:, resource_id:)
          @monitor.synchronize do
            rec = @runs[key]
            return unless rec

            rec.wait_count += 1
            rec.waiting_since = now if rec.wait_count == 1
            rec.wait_kind = wait_kind
            rec.wait_resource_id = resource_id
          end
        end

        # Mark the run as active (decrement the wait counter). Floors at zero so
        # unmatched +mark_active+ calls never underflow.
        def mark_active(key, now:)
          @monitor.synchronize do
            rec = @runs[key]
            return unless rec

            rec.wait_count = [rec.wait_count - 1, 0].max
            if rec.wait_count.zero?
              rec.waiting_since = nil
              rec.wait_kind = nil
              rec.wait_resource_id = nil
            end
          end
        end

        # Select an eviction candidate, preferring TTL-eligible candidates first,
        # then capacity-pressure candidates (only when above the waiting
        # capacity threshold).
        #
        # @return [String, nil] The action key of the chosen candidate, or nil
        def select_eviction_candidate(now:, durable_slots:, reserve_slots:, min_wait_for_capacity_eviction:)
          @monitor.synchronize do
            waiting = @runs.values.select do |r|
              r.waiting? && !r.eviction_policy.nil?
            end
            return nil if waiting.empty?

            ttl_eligible = waiting.select do |r|
              policy = r.eviction_policy
              policy&.ttl && r.waiting_since && (now - r.waiting_since) >= policy.ttl
            end

            unless ttl_eligible.empty?
              chosen = ttl_eligible.min_by do |r|
                [r.eviction_policy ? r.eviction_policy.priority : 0, r.waiting_since || now]
              end
              ttl = chosen.eviction_policy&.ttl
              chosen.eviction_reason = DurableEvictionCache.build_eviction_reason(
                EvictionCause::TTL_EXCEEDED, chosen, ttl: ttl,
              )
              return chosen.key
            end

            return nil unless capacity_pressure?(durable_slots, reserve_slots, waiting.length)

            capacity_candidates = waiting.select do |r|
              r.eviction_policy&.allow_capacity_eviction &&
                r.waiting_since &&
                (now - r.waiting_since) >= min_wait_for_capacity_eviction
            end
            return nil if capacity_candidates.empty?

            chosen = capacity_candidates.min_by do |r|
              [r.eviction_policy ? r.eviction_policy.priority : 0, r.waiting_since || now]
            end
            chosen.eviction_reason = DurableEvictionCache.build_eviction_reason(
              EvictionCause::CAPACITY_PRESSURE, chosen,
            )
            chosen.key
          end
        end

        # Build a human-readable eviction reason string.
        def self.build_eviction_reason(cause, rec, ttl: nil)
          wait_desc = rec.wait_kind || "unknown"
          wait_desc = "#{wait_desc}(#{rec.wait_resource_id})" if rec.wait_resource_id

          case cause
          when EvictionCause::TTL_EXCEEDED
            ttl_str = ttl ? " (#{ttl}s)" : ""
            "Wait TTL#{ttl_str} exceeded while waiting on #{wait_desc}"
          when EvictionCause::CAPACITY_PRESSURE
            "Worker at capacity while waiting on #{wait_desc}"
          when EvictionCause::WORKER_SHUTDOWN
            "Worker shutdown while waiting on #{wait_desc}"
          else
            raise ArgumentError, "Unknown eviction cause: #{cause}"
          end
        end

        private

        def capacity_pressure?(durable_slots, reserve_slots, waiting_count)
          return false if durable_slots <= 0

          max_waiting = durable_slots - reserve_slots
          return false if max_waiting <= 0

          waiting_count >= max_waiting
        end
      end
    end
  end
end
